File: 백준/레벨Gold IV/1753. 최단경로/최단경로.cc
#include <bits/stdc++.h>

using namespace std;
using pii = pair<int, int>;

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  // 인접리스트 초기화
  int V, E, S;
  cin >> V >> E >> S;

  vector<vector<pii>> graph(V + 1);

  while (E--) {
    int u, v, w;
    cin >> u >> v >> w;

    graph[u].emplace_back(v, w);
  }

  // 거리배열 초기화
  int INF = INT_MAX;
  vector<int> distances(V + 1, INF);
  distances[S] = 0;

  // 다익스트라
  priority_queue<pii, vector<pii>, greater<>> pq;
  pq.emplace(0, S);

  while (!pq.empty()) {
    auto [currDist, u] = pq.top();
    pq.pop();

    if (currDist > distances[u]) continue;

    for (auto [v, w] : graph[u]) {
      int nextDist = currDist + w;

      if (nextDist < distances[v]) {
        distances[v] = nextDist;
        pq.emplace(nextDist, v);
      }
    }
  }

  // 출력 처리
  for (size_t i = 1; i < distances.size(); i++) {
    if (distances[i] == INF) {
      cout << "INF";
    } else {
      cout << distances[i];
    }
    cout << "\n";
  }
}  

File: 백준/레벨Gold IV/1753. 최단경로/최단경로.py
from sys import stdin
from heapq import heappush, heappop

input = stdin.readline

# 인접리스트 초기화
V, E = map(int, input().split())
S = int(input().rstrip())
graph = [[] for _ in range(V + 1)]

for _ in range(E):
    u, v, w = map(int, input().split())
    graph[u].append((v, w))

# 거리배열 초기화
INF = float('inf')
distances = [INF] * (V + 1)
distances[S] = 0

# 다익스트라
heap = []
heappush(heap, (0, S))

while heap:
    curr_dist, u = heappop(heap)

    if (curr_dist > distances[u]):
        continue
    
    for v, w in graph[u]:
        next_dist = curr_dist + w

        if next_dist < distances[v]:
            distances[v] = next_dist
            heappush(heap, (next_dist, v))

# 출력 처리
for i in range(1, len(distances)):
    if distances[i] == INF:
        print("INF")
    else:
        print(distances[i])